### PR TITLE
Consolidate TeachingSubject presentation

### PR DESCRIPTION
**Pull request by ethax-ross**
### Trello card

[Trello-4591](https://trello.com/c/UQdgvC0I/4591-consolidate-and-streamline-teaching-subjects-on-git-funnels)

### Context

We currently retrieve teaching subjects from the API in a multitude of different ways. We want to be consistent in how we retrieve subjects and also in how we filter/display them.

### Changes proposed in this pull request

- Consolidate TeachingSubject presentation

Centralise the teaching subject queries to the `Crm::TeachingSubject` model and update all references to go through this.

### CODE REVIEW

File: spec/helpers/welcome_helper_spec.rb [hunk 1]
```diff
@@ -89,7 +89,7 @@
 
     context "when stored in the mailinglist session store" do
       context "when the subject name is a common noun" do
-        %i[art_and_design biology chemistry general_science languages_other maths physics_with_maths physics].each do |subject_key|
```
Comment (ethax-ross) [inline spec/helpers/welcome_helper_spec.rb:92]: Why are we removing these?

File: app/models/crm/teaching_subject.rb [hunk 2]
```diff
@@ -51,27 +23,35 @@ def lookup_by_keys(*keys)
...
+    def all_hash
+      all.to_h { |item| [item.value, item.id] }
+    end
+
+    def all_without_primary
+      all_hash.reject { |_h, s| s == PRIMARY }
+    end
+
+    def all
+      GetIntoTeachingApiClient::LookupItemsApi.new.get_teaching_subjects.reject { |subject| IGNORED.value?(subject.id) }.each { |subject| if subject && RENAMED[subject.value].present? then subject.value = RENAMED[subject.value]; end; }.sort_by { |f| [f.value] }
```
Comment (ethax-ross) [inline app/models/crm/teaching_subject.rb:54]: I think this needs to be simplified its quite hard to follow. We could condense it to something along the lines of:

```
def all
  GetIntoTeachingApiClient::LookupItemsApi.new.get_teaching_subjects
    .sort_by(&:value)
    .reject(&method(:ignored?))
    .each(&method(:rename))
end

private

def rename(subject)
  subject.value = RENAMED[subject.value] || subject.value
end

def ignored?(subject)
  IGNORED.value?(subject.id)
end
```

As a follow-up PR I think we should look at removing the `all_hash` from this class, but that would have many knock-on changes so it can be done separately

**Review by ethax-ross — COMMENTED** [review]
Looking good, just a couple of questions and a suggestion

File: spec/features/event_wizard_spec.rb [hunk 3]
```diff
@@ -336,7 +336,7 @@ def personalised_attributes
     {
       degree_status_id: 222_750_000,
       consideration_journey_stage_id: 222_750_000,
-      preferred_teaching_subject_id: Crm::TeachingSubject.lookup_by_key(:art),
+      preferred_teaching_subject_id: Crm::TeachingSubject.lookup_by_key(:art_and_design),
```
[the remark this answers is not in the document]
Comment (plisovin) [reply spec/features/event_wizard_spec.rb:339]: ok I will check, I think I've made wrong assumption that this is just an example subject used in this test and as the subject got removed then a next one could be picked up from the list. I will investigate it further

File: spec/helpers/welcome_helper_spec.rb (hunk 1 above)
Comment (plisovin) [reply spec/helpers/welcome_helper_spec.rb:92]: Same reason as above.

File: app/models/crm/teaching_subject.rb (hunk 2 above)
Comment (plisovin) [reply app/models/crm/teaching_subject.rb:54]: yeah, good idea, I just didn't look through as just fixed it so rubocop doesn't complain

**Review by plisovin — APPROVED** [review]

**Comment by gemmadallmandfe** [comment]
Looks good @plisovin - just a couple of minor things:

* Can you please add `General science` back in again but rename it as `Science` (my fault - sorry!)

* Can the subjects be displayed alphabetically? `Communication and media studies` is appearing where `Media studies` previously did but this is no longer in alphabetical order
NB this will also be the case when `General science` is renamed as `Science`

image

* I'm presuming that only the adviser funnel on GIT will be updated with the modified subject list, and not the original adviser funnel? I don't think that is an issue particularly - just want to be clear


I will need to let the adviser team know of the changes so please don't merge this in the meantime :-)

File: spec/helpers/welcome_helper_spec.rb [hunk 4]
```diff
@@ -89,7 +89,7 @@
...
+        %i[art_and_design biology chemistry languages_other maths physics].each do |subject_key|
```
Comment (gemmadallmandfe) [inline spec/helpers/welcome_helper_spec.rb:92]: @plisovin Do we need to add `science` in here now that we've added it back in as a subject?

**Review by gemmadallmandfe — APPROVED** [review]
Looks good to me - happy for this to be merged tomorrow @plisovin

One thing I've noticed is that a couple of questions in the adviser funnel have an additional option of 'Other'

'/teacher-training-adviser/sign_up/subject_taught'
Which main subject did you previously teach?

`/teacher-training-adviser/sign_up/subject_like_to_teach`
Which subject would you like to teach if you return to teaching?

'/teacher-training-adviser/sign_up/what_subject_degree`
What subject is your degree?

We may need to introduce this 'Other' option back in again for these specific questions - however, I will raise a separate ticket for this rather than hold up this PR

We then have work planned in later this summer to address degree subject properly!
